File: landrop-main/receiver/android/app/src/test/java/com/landrop/receiver/protocol/NetworkingTest.kt
package com.landrop.receiver.protocol

import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test

class NetworkingTest {

    private val discoveryInfo = DiscoveryInfo(
        service = DISCOVERY_SERVICE_TYPE,
        hostname = "receiver-device.local",
        port = 45821,
        protocolVersion = 1,
        deviceName = "Android Device",
    )

    @Test
    fun networkingFlowReachesSessionEstablished() {
        // discovery (protocol §4)
        val registry = DiscoveryRegistry()
        registry.upsert(discoveryInfo)
        assertEquals(1, registry.size)
        assertTrue(isDiscoverySupportedVersion(discoveryInfo))

        // connection (protocol §5 steps 1-2)
        val connection = ConnectionAttempt()
        assertEquals(DeliveryEvent.CONNECT_INITIATED, connection.begin())
        assertEquals(null, connection.complete(ConnectionOutcome.ESTABLISHED))

        // TLS 1.3 on the open transport (steps 3-4)
        val channel = SecureChannelAttempt()
        channel.begin()
        assertEquals(DeliveryEvent.SECURE_CHANNEL_ESTABLISHED, channel.complete(SecureChannelOutcome.ESTABLISHED))

        // negotiation (step 5) + session establishment (steps 6-7)
        assertTrue(supportsProtocolVersion(1))
        val est = establishSession(
            SessionDraft(
                sessionId = "sess_01JTEST",
                protocolVersion = 1,
                sender = SenderProfile("ISHAQ CYBERTECH"),
                receiver = DeviceInfo("Test Device"),
                lifetime = Lifetime(0, 1000),
            ),
            EstablishmentPrerequisites(transportEstablished = true, secureChannelEstablished = true),
        ) as SessionEstablishmentResult.Ok
        assertEquals(DeliveryState.DISCOVERING, est.session.state)
        assertEquals(DeliveryEvent.SESSION_ESTABLISHED, sessionEstablishedEvent())

        // delivery machine mapping (protocol §5 steps 1-7)
        var state = DeliveryStateMachine.initialDeliveryState
        val steps = listOf(
            DeliveryEvent.DEVICE_FOUND to DeliveryState.AVAILABLE,
            DeliveryEvent.CONNECT_INITIATED to DeliveryState.CONNECTING,
            DeliveryEvent.SECURE_CHANNEL_ESTABLISHED to DeliveryState.SECURE_CHANNEL,
            DeliveryEvent.SESSION_ESTABLISHED to DeliveryState.SESSION_ESTABLISHED,
        )
        for ((event, expected) in steps) {
            assertEquals(expected, (DeliveryStateMachine.transition(state, event) as DeliveryTransition.Accepted).to)
            state = expected
        }
        assertEquals(DeliveryState.SESSION_ESTABLISHED, state)
    }

    @Test
    fun refusedConnectionFlowIsConnectionLost() {
        val connection = ConnectionAttempt()
        connection.begin()
        assertEquals(DeliveryEvent.CONNECTION_LOST, connection.complete(ConnectionOutcome.REFUSED))
        assertEquals(ConnectionPhase.FAILED, connection.phase)
        assertEquals(
            listOf(DeliveryEvent.CONNECT_INITIATED, DeliveryEvent.CONNECTION_LOST),
            connection.events(),
        )
        assertEquals(
            DeliveryState.FAILED,
            (DeliveryStateMachine.transition(DeliveryState.CONNECTING, DeliveryEvent.CONNECTION_LOST) as DeliveryTransition.Accepted).to,
        )
    }

    @Test
    fun connectionLostDrivesReachedStagesToFailed() {
        for (state in listOf(
            DeliveryState.CONNECTING,
            DeliveryState.SECURE_CHANNEL,
            DeliveryState.SESSION_ESTABLISHED,
        )) {
            assertEquals(
                DeliveryState.FAILED,
                (DeliveryStateMachine.transition(state, DeliveryEvent.CONNECTION_LOST) as DeliveryTransition.Accepted).to,
            )
        }
        assertTrue(DeliveryStateMachine.isDeliveryTerminal(DeliveryState.FAILED))
        assertTrue(
            DeliveryStateMachine.transition(DeliveryState.FAILED, DeliveryEvent.CONNECTION_LOST)
                is DeliveryTransition.Rejected,
        )
    }

    @Test
    fun networkingLayersMapOnlyIntoDefinedDeliveryEvents() {
        val emitted = listOf(
            DeliveryEvent.DEVICE_FOUND,
            DeliveryEvent.CONNECT_INITIATED,
            DeliveryEvent.SECURE_CHANNEL_ESTABLISHED,
            DeliveryEvent.SESSION_ESTABLISHED,
            DeliveryEvent.CONNECTION_LOST,
        )
        val defined = DeliveryEvent.values().toList()
        assertTrue(defined.containsAll(emitted))
    }
}

File: landrop-main/sender/desktop/src/protocol/networking.test.ts
import { describe, expect, it } from "vitest";

import { ConnectionAttempt } from "./connection";
import { SecureChannelAttempt } from "./secure-channel";
import { establishSession, sessionEstablishedEvent, supportsProtocolVersion } from "./session-establishment";
import { parseDiscoveryInfo, createDiscoveryRegistry } from "./discovery";
import { createSession } from "./session";
import { INITIAL_DELIVERY_STATE } from "./state-machine";
import { createInMemoryDuplex } from "./transport";
import { JsonMessageStream, encodeJsonFrame } from "./message-stream";
import type { JsonMessageResult } from "./message-stream";
import { transition } from "./state-machine";
import type { DeliveryEvent } from "./state-machine";
import type { ControlMessage, DeliveryState } from "./types";

const fixtures: Record<string, unknown> = require(
  "../../../../protocol/fixtures/model-cases.json",
);

const discovery = fixtures.discovery as {
  peer: { service: string; hostname: string; port: number; protocol_version: number; device_name: string };
};

const deliveryEventMapping = fixtures.delivery_event_mapping as {
  layer: string;
  condition: string;
  event: DeliveryEvent;
}[];

const VALID_REQUEST: ControlMessage = {
  type: "delivery_request",
  protocol_version: 1,
  request_id: "req_01JTEST",
  session_id: "sess_01JTEST",
  presentation: { mode: "GUI" },
  sender: { display_name: "ISHAQ CYBERTECH" },
  application: {
    name: "My Application",
    version: "1.4.2",
    description: "A local test application.",
    package_name: "com.example.application",
    size_bytes: 26004608,
    sha256:
      "0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef",
  },
};

const SESSION_DRAFT = {
  session_id: "sess_01JTEST",
  protocol_version: 1,
  sender: { display_name: "ISHAQ CYBERTECH" },
  receiver: { device_name: "Test Device" },
  lifetime: { created_at_ms: 0, expires_at_ms: 1000 },
};

const establish = () => {
  const established = establishSession(SESSION_DRAFT, {
    transportEstablished: true,
    secureChannelEstablished: true,
  });
  if (!established.ok) throw new Error("unexpected establishment failure");
  return established;
};

const RUN_STATES = ["CONNECTING", "SECURE_CHANNEL", "SESSION_ESTABLISHED"] as const;

describe("phase 05 — networking integration over Phase 02-04 foundations", () => {
  it("follows the fixture flow: discover -> connect -> TLS -> negotiate -> session", () => {
    // Discovery yields a candidate device (protocol §4).
    expect(supportsProtocolVersion(1)).toBe(true);
    const parsed = parseDiscoveryInfo(discovery.peer);
    expect(parsed.ok).toBe(true);
    const registry = createDiscoveryRegistry();
    if (parsed.ok) registry.upsert(parsed.info);
    expect(registry.list()).toHaveLength(1);

    // Delivery machine mapping (protocol §5 steps 1-7).
    const steps: [DeliveryState, DeliveryEvent, DeliveryState][] = [
      ["DISCOVERING", "device_found", "AVAILABLE"],
      ["AVAILABLE", "connect_initiated", "CONNECTING"],
      ["CONNECTING", "secure_channel_established", "SECURE_CHANNEL"],
      ["SECURE_CHANNEL", "session_established", "SESSION_ESTABLISHED"],
    ];
    let current: DeliveryState = "DISCOVERING";
    for (const [from, event, to] of steps) {
      const next = transition(current, event);
      expect(next.ok, `${from} --${event}--> ${to}`).toBe(true);
      if (next.ok) expect(next.to).toBe(to);
      current = to;
    }
    expect(current).toBe("SESSION_ESTABLISHED");
  });

  it("the fixture flow table is consistent with the state machine", () => {
    expect(INITIAL_DELIVERY_STATE).toBe("DISCOVERING");
    const expects: [DeliveryState, DeliveryEvent, DeliveryState][] = [
      ["DISCOVERING", "device_found", "AVAILABLE"],
      ["AVAILABLE", "connect_initiated", "CONNECTING"],
      ["CONNECTING", "secure_channel_established", "SECURE_CHANNEL"],
      ["SECURE_CHANNEL", "session_established", "SESSION_ESTABLISHED"],
    ];
    const wanted = new Map(expects.map(([from, event]) => [`${from}--${event}`, event]));
    expect(wanted.size).toBe(4);
    for (const [from, event, to] of expects) {
      const next = transition(from, event);
      expect(next.ok, `${from} --${event}-->`).toBe(true);
      if (next.ok) expect(next.to).toBe(to);
    }
    expect(deliveryEventMapping.length).toBeGreaterThanOrEqual(5);
    expect(deliveryEventMapping.map((r) => r.event)).toContain("connection_lost");
  });

  it("a lost connection pushes the flow to FAILED (protocol §27, §30)", () => {
    for (const from of RUN_STATES) {
      const lost = transition(from, "connection_lost");
      expect(lost.ok, from).toBe(true);
      if (lost.ok) expect(lost.to).toBe("FAILED");
      expect(transition("FAILED", "connection_lost").ok).toBe(false); // terminal
    }
  });

  it("a refused connect produces exactly [connect_initiated, connection_lost]", () => {
    const connection = new ConnectionAttempt();
    const events: DeliveryEvent[] = [];
    const begin = connection.begin();
    if (begin !== null) events.push(begin);
    const fail = connection.complete("refused");
    if (fail !== null) events.push(fail);
    expect(events).toEqual(["connect_initiated", "connection_lost"]);
    expect(connection.currentPhase).toBe("failed");
    expect(connection.lastFailure?.kind).toBe("IO_ERROR");
  });

  it("full handshake session on a deterministic channel", () => {
    const connection = new ConnectionAttempt();
    const channel = new SecureChannelAttempt();
    const flow: DeliveryEvent[] = [];
    const push = (event: DeliveryEvent | null): void => {
      if (event !== null) flow.push(event);
    };
    push(connection.begin());
    push(connection.complete("established"));
    push(channel.begin());
    push(channel.complete("established"));
    const shouldBeSessionEstablished = establish();
    expect(shouldBeSessionEstablished.session.ok).toBe(true);
    push(sessionEstablishedEvent());
    push(connection.close());
    expect(flow).toEqual([
      "connect_initiated",
      "secure_channel_established",
      "session_established",
      "connection_lost",
    ]);
    // Determinism re-run.
    expect(flow).toEqual([
      "connect_initiated",
      "secure_channel_established",
      "session_established",
      "connection_lost",
    ]);
  });

  it("a closed TLS connection is classified connection_lost, not a session", () => {
    const connection = new ConnectionAttempt();
    const channel = new SecureChannelAttempt();
    connection.begin();
    connection.complete("established");
    channel.begin();
    expect(channel.complete("closed")).toBe("connection_lost");
    expect(connection.currentPhase).toBe("established"); // transport must close separately
    const loss = connection.close();
    expect(loss).toBe("connection_lost");
  });

  it("a delivery request travels codec -> frame -> duplex transport intact", () => {
    const duplex = createInMemoryDuplex();
    const encoded = encodeJsonFrame(VALID_REQUEST);
    expect(encoded.ok).toBe(true);
    if (encoded.ok) {
      expect(duplex.left.connect().ok).toBe(true);
      expect(duplex.right.connect().ok).toBe(true);
      expect(duplex.left.write(encoded.bytes).ok).toBe(true);

      // Pump transport.read into the message stream (architecture §29).
      const stream = new JsonMessageStream();
      let message: JsonMessageResult = stream.readMessage();
      if (message.outcome === "incomplete") {
        const read = duplex.right.read(4096);
        expect(read.kind).toBe("data");
        if (read.kind === "data") stream.feed(read.bytes);
        message = stream.readMessage();
      }
      expect(message.outcome).toBe("message");
      if (message.outcome === "message") {
        expect(message.message.type).toBe("delivery_request");
        expect(stream.bufferedBytes).toBe(0);
      }
    }
  });

  it("no messages or sessions can be forged after connection loss", () => {
    const session = createSession(SESSION_DRAFT);
    expect(session.ok).toBe(true);
    // After SESSION_ESTABLISHED -(connection_lost)> FAILED, reactivation is
    // unreachable: secure_channel_established/session_established are invalid.
    const afterLost = transition("FAILED", "session_established");
    expect(afterLost.ok).toBe(false);
  });
});